grove/measurements/estimation.py

```diff
@@ -0,0 +1,161 @@
+from pyquil.paulis import PauliSum, PauliTerm, commuting_sets, sI
+from pyquil.quil import Program
+from pyquil.gates import RY, RX
+
+
+class CommutationError(Exception):
+    pass
+
+
+def remove_imaginary_terms(pauli_sums):
+    """
+    Remove the imaginary components of the Pauli sum term
+
+    :param pauli_sums:
+    :return:
+    """
+    if not isinstance(pauli_sums, PauliSum):
+        raise TypeError("not a pauli sum. please give me one")
+    new_term = sI(0) * 0.0
+    for term in pauli_sums:
+        if (not np.isclose(term.coefficient.real, 0.0) and
+                np.isclose(term.coefficient.imag, 0.0)):
+            new_term += term
+
+    return new_term
+
+
+def get_rotation_program(pauli_term):
+    """
+    Generate a rotation program so that the pauli term is diagoanl
+
+    :param pauli_term:
+    :return: rotaiton program
+    :rtype: Program
+    """
+    meas_basis_change = Program()
+    for index, gate in pauli_term:
+        if gate == 'X':
+            meas_basis_change.inst(RY(-np.pi / 2, index))
+        elif gate == 'Y':
+            meas_basis_change.inst(RX(np.pi / 2, index))
+        elif gate == 'Z':
+            pass
+        else:
+            raise ValueError()
+
+    return meas_basis_change
+
+
+def get_parity(pauli_terms, bitstring_results):
+    """
+    Calculate the parity at each of the marked qubits in pauli_terms
+
+    Each list of bits in the bitstring_results is the maximum weight pauli term
+    in pauli_terms.  The bits in each element of bitstring_results correspond
+    to a numerical ordering of the maximal weight element of pauli_terms.
+
+    An example:
+
+    A maximum weight pauli term is Z1Z5Z6.
+
+    each element of bitstring_results is something in
+    :math:`\{0, 1\}^{\otimes 3}` where the first bit corresponds to the measured
+    value of qubit 1, the second bit corresponds to the measured value of qubit
+    5, and the third bit corresponds to the measured value of qubit 6.
+
+    :param pauli_terms:
+    :param bitstring_results:
+    :return:
+    """
+    max_weight_pauli_index = np.argmax(
+        [len(term.get_qubits()) for term in pauli_terms])
+    active_qubit_indices = sorted(
+        pauli_terms[max_weight_pauli_index]._ops.keys())
+    index_mapper = dict(zip(active_qubit_indices,
+                            range(len(active_qubit_indices))))
+
+    results = np.zeros((len(pauli_terms), len(bitstring_results)))
+
+    # convert to array so we can fancy index into it later.
+    # list() is needed to cast because we don't really want a map object
+    bitstring_results = list(map(np.array, bitstring_results))
+    for row_idx, term in enumerate(pauli_terms):
+        memory_index = np.array(list(map(lambda x: index_mapper[x],
+                                         sorted(term.get_qubits()))))
+
+        results[row_idx, :] = [-2 * (sum(x[memory_index]) % 2) + \
+                               1 for x in bitstring_results]
+    return results
```

Comment:
So, depending on your answer re the docstring, i.e., what exactly this function does, I have a hunch this code could be made easier to follow

Reply:
Happy to hear ways to improve it!